setup.py

```diff
@@ -80,6 +86,89 @@ class Components:
 else:
     from setuptools.command.build_ext import build_ext
 
+music_home = ""
+if Components.MUSIC:
+    """
+    If music is not installed, try to install it. In principle, a dynamic
+    build requires only music include files. But demand a <full> installation
+    where music is in the PATH and can find include and lib. If not installed,
+    give up with Components.MUSIC = False and go on. Otherwise set some
+    variables for cmake.
+    """
+
+    def run(cmd, env, err=True):
+        result = subprocess.run(
+            cmd, shell=True, env=env, capture_output=True, text=True
+        )
+        if err:
+            result.check_returncode()
+        return result
+
+    def is_music_installed(env, pr_err=False):
+        # returns home for music or "" if music not in PATH
+        music_home = ""
+        try:
+            result = run("which music", env)
+            music_home = "/".join(result.stdout.strip().split("/")[:-2])
+            from os.path import exists
+
+            # does the library exist?
+            name = ""
+            for suffix in ["so", "dylib", "dll", None]:
+                name = music_home + "/lib/libmusic." + suffix
+                if exists(name):
+                    break
+            if name == "":
+                raise FileNotFoundError("no libmusic in {}/lib".format(musicHome))
+            # does include exist
+            name = music_home + "/include/music.hh"
+            if not exists(name):
+                raise FileNotFoundError(name + " does not exist")
+        except Exception as e:
+            if pr_err:
+                print(e)
+            music_home = ""
+        return music_home
+
+    def install_mpi4py(myenv):
+        installed = True
+        try:
+            import mpi4py
+        except:
+            installed = False
+        if not installed:
+            run(sys.executable + " -m pip install mpi4py", myenv)
+
+    def install_music():
+        myenv = os.environ.copy()
+        if not is_music_installed(myenv) and os.path.exists("/nrnwheel"):
+            print("Attempt to install MUSIC")
+            try:
+                install_mpi4py(myenv)
+            except:
+                return ""
+            cmd = r"""curl -L -o MUSIC.zip https://github.com/INCF/MUSIC/archive/refs/heads/switch-to-MPI-C-interface.zip \
+                     && unzip MUSIC.zip \
+                     && mv MUSIC-switch-to-MPI-C-interface MUSIC \
+                     && cd MUSIC \
+                     && ./autogen.sh \
+                     && ./configure --prefix=/nrnwheel/MUSIC --with-python-sys-prefix --disable-anysource \
+                     && make -j install
+                  """
+            myenv["PATH"] = myenv["PATH"] + ":/nrnwheel/MUSIC"
+            result = run(cmd, myenv, err=False)
+            if result.returncode:
+                print(result)
+                return ""
+        return is_music_installed(myenv, pr_err=True)
```

Comment:
as part of setup.py, we definitely shouldn't do this :) i.e. **source-based installation of MUSIC** shouldn't be triggered by neuron's setup.py.  If there would have been a wheel for MUSIC, we could have added as a dependent package. 

I assume it's ok to rely on the user to make MUSIC available?

If your question is about having MUSIC installation available for building wheel then we install that in Dockerfile (as you did) and [the update build_wheel.sh](https://github.com/neuronsimulator/nrn/blob/ce1caabb682afe1f8ca1a9385e88b78fab872e93/packaging/python/build_wheels.bash#L108) to pass the necessary CMake prefix option.

Reply:
Ok. My mistake. I did not intend for it to be a source based installation when the user installs a wheel. It was intended to just allow a wheel to be constructed with cmake args ```-DNRN_ENABLE_MUSIC=ON``` successfully. (actually the only requirement for a wheel build (I think) is to find a valid music.hh file, but I'm confused as to whether the neuronmusic module needs a full installation of music at wheel build time. Perhaps a bit more experimentation is in order.  Up to now, I've just been trying to get a wheel actually built and tested with and without music on the user machine.  Also at the moment I'm allowing wheel construction to continue if the music part fails. That is probably a mistake as well.